linux-v4l2: Add option to close video device when not visible

julijane opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

### Description
This change adds the ability to configure the source to close the video device when hidden.

### Motivation and Context
Currently there is no proper way to "restart"/"reset" a video input device except changing configuration - and since a recent PR was merged even that only happens when you really force it. Adding a possibility to have the video device closed when the input is hidden adds an elegant way to enable to user to do that e.g. to try to fix an issue or just to free the video device (i.e. when it is configured to the wrong source and you want to activate it in a different source).

Closes #2775

### How Has This Been Tested?

As it is a very simple change I compiled it, started OBS, enabled the feature and tried many hide/show and watched console output for any issues. All worked fine.

### Types of changes
New feature

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

WizardCM commented on plugins/linux-v4l2/data/locale/en-US.ini:
```diff
@@ -13,3 +13,4 @@ ColorRange.Default="Default"
 ColorRange.Partial="Partial"
 ColorRange.Full="Full"
 CameraCtrls="Camera Controls"
+CloseOnHide="Close device when not visible"
```
In win-dshow, We use the string "Deactivate when not showing". I think it makes sense to use the same string here.

Additionally, you should only enclude the English text. Other languages are fetched using Crowdin.

kkartaltepe commented on plugins/linux-v4l2/v4l2-input.c:
```diff
@@ -796,6 +797,9 @@ static obs_properties_t *v4l2_properties(void *vptr)
 	obs_properties_add_bool(props, "buffering",
 				obs_module_text("UseBuffering"));
 
+	obs_properties_add_bool(props, "close_on_hide",
```
Use the string `deactivate_when_not_showing` to mirror windows dshow devices.

julijane replied:
@kkartaltepe Thank you. There is another issue and that is the spelling of variable names. A recently merged PR added two other boolean variables to the data struct and there CamelCase was used. Which I now have also used for my boolean. However all other variables in the struct have kebab_case. So shouldn't I use kebab_case too and the other two variables be changed (in a different patch/pr) also?

kkartaltepe commented:
This doesnt appear to work correctly.

Adding two sources with the new setting enabled to a scene and setting both to disabled seems to be broken. I got one cam with a frozen frame and the other working when playing with it. after restarting OBS one of the sources was never able to capture the camera.

julijane commented:
@kkartaltepe Thank you for your testing and feedback. The issue was caused by not keeping proper track if the source is shown or not. I have added code for that and tested it, it appears to me that the issue is now fixed. Please try again.

Fenrirthviti commented:
Please squash all your commits and update them to have the proper commit message per our contribution guidelines.

julijane commented:
I have changed the PR accordingly.

kkartaltepe commented:
> @kkartaltepe Thank you for your testing and feedback. The issue was caused by not keeping proper track if the source is shown or not. I have added code for that and tested it, it appears to me that the issue is now fixed. Please try again.

Devices look like they successfully reinit now. However the issue with last frame hanging on when camera cannot init still seems to be there. I will try and write better reproduction steps later, but its mostly playing with the same scene with two of the same capture device.

kkartaltepe commented:
On my system the old frame is held when you do this

1. Start obs with both inputs disabled.

1. Enable and disable input 1 allowing it to initialize.

1. Enable and disable input 2 allowing it to initialize.

1. Enable input 1

1. Enable input 2 (Here 2 fails to initialize and shouldnt render anything, instead it renders the last frame from before it was hidden in step 3)

julijane commented:
@kkartaltepe This should be fixed by the updated PR.

julijane commented:
@WizardCM I do not understand the project structure and who is responsible for what, but maybe you can help. It would be really nice to have this merged. What can I do?

Fenrirthviti commented:
Finding testers to confirm this works as expected across multiple different systems/distributions would help with confidence this doesn't cause issues, but mainly it is just waiting until we have time to do a proper review.  Jim is pretty much the only person who does final code approvals currently, and his time is limited.

kkartaltepe commented:
I'll try to test this within the next few days.

julijane commented:
@Fenrirthviti @kkartaltepe Thank you very much for getting back to me. I do not want to put pressure upon anyone. But I intend to work on some more fundamental changes and it would change the same code, so it is a bit of an issue if I don't know if this will be merged or not :) What I plan would be a great improvement (short version: The current code for device "matching" is really not good and causes lots of issues with my 4 cam setup and I'm so sick and tired of it that I want to introduce a more robust solution which is in the works).

kkartaltepe commented:
Rebased on master and it appears to race when switching scenes. In both scenes the device is set to deactivate when not showing.

```
info: User switched to scene 'Scene 1'
info: v4l2-input: Start capture from /dev/video0
info: v4l2-input: Input: 0
info: v4l2-input: Resolution: 800x448
info: v4l2-input: Pixelformat: BGR3
info: v4l2-input: Linesize: 2400 Bytes
info: v4l2-input: Framerate: 30.00 fps
info: User switched to scene 'Scene 2'
info: v4l2-input: Start capture from /dev/video0
error: v4l2-input: Unable to set input 0
error: v4l2-input: Initialization failed
info: v4l2-input: Stopped capture after 51 frames
```

Seems the device is not closed until after the new scene attempts to start capture.

kkartaltepe commented:
It also ended up somehow failing to initialize the devices with settings that work on master (tested after being rebased onto master). Its not quite clear what I changed in my configuration that caused it to start failing. But now just deleting all sources and trying to add the device back fails.

julijane commented:
@kkartaltepe Trying to understand what you did. Do I get it right that you added the same cam as two different souces, one in each scene?

jp9000 commented:
(Just wanted to chime in and say thank you for doing more thorough testing, I just haven't had time to go over PRs right now)

kkartaltepe commented:
> @kkartaltepe Thank you for testing. Trying to understand what you did. Do I get it right that you added the same cam as two different souces, one in each scene?

Right.

Im not sure what triggered the second issue unfortunately.

WizardCM commented:
Heads up: this PR has merge conflicts.

julijane commented:
@WizardCM Thanks. Maybe I shall close this PR for now, this could need some rework I guess. The whole v4l2 implementation could, I would love better device matching, based on busid (where available i.e. Linux). I actually worked on that back then too but found it difficult without breaking backwards compatibility and due to the fact that the code is also used on bsd where busid is not available (i believe).
